File: solutions/csharp/perfect-numbers/2/PerfectNumbers.cs
public enum Classification
{
    Deficient = -1,
    Perfect = 0,
    Abundant = 1
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(number));
        }

        return (Classification)AliquotSum(number).CompareTo(number);
    }

    private static int AliquotSum(int number)
    {
        if (number == 1) return 0;

        return Enumerable
            .Range(2, (int)Math.Sqrt(number) - 1)
            .Where(IsMultipleOf(number))
            .SelectMany(FactorPair(number))
            .Sum() + 1;
    }

    private static Func<int, bool> IsMultipleOf(int number) => factor => number % factor == 0;

    private static Func<int, IEnumerable<int>> FactorPair(int number) => factor =>
    {
        var pair = number / factor;

        return pair == factor ? [factor] : [factor, pair];
    };
}

File: solutions/scala/perfect-numbers/2/src/main/scala/PerfectNumbers.scala
enum NumberType {
  case Perfect, Abundant, Deficient
}

object PerfectNumbers {
  private val InvalidNumber = "Classification is only possible for natural numbers."

  def classify(number: Int): Either[String, NumberType] =
    if number < 1 then Left(InvalidNumber)
    else Right(number.compareTo(aliquotSum(number)) match {
      case 0 => NumberType.Perfect
      case 1 => NumberType.Deficient
      case -1 => NumberType.Abundant
    })

  private def aliquotSum(number: Int): Int =
    if number == 1 then 0
    else
      // The collection version below is concise, but timed out in the online runner:
      // val pairedFactors = (2 to root).filter(number % _ == 0).flatMap(factor => pairedFactor(number, factor))
      // (1 +: pairedFactors).sum
      //
      // Keep the sqrt bound, but avoid per-factor collection allocations.
      val root = math.sqrt(number).toInt
      var sum = 1
      var factor = 2

      while factor <= root do
        if number % factor == 0 then
          val pair = number / factor
          sum += factor

          if pair != factor then
            sum += pair

        factor += 1

      sum
}
